From 752e7838b9e4683b5ae5a6d4c8418e2dbc54486c Mon Sep 17 00:00:00 2001
From: 5656565566 <contact@home56.top>
Date: Fri, 14 Aug 2026 02:12:00 +0800
Subject: [PATCH] frameworks: separate Floral public and loader ABI views

Expose the configured Floral application ABI view through Build while retaining
build-owned ABI lists for zygote and platform internals. PackageManager keeps an
APK on the public ARM family whenever any base or split APK contains ARM code.

Change-Id: I58859c26b8b3559f7848cc51e3c68f7dfb809822
---
 .../com/android/commands/am/Instrument.java   |  2 +-
 core/java/android/os/AppZygote.java           |  2 +-
 core/java/android/os/Build.java               | 35 ++++++++++--
 core/java/android/os/ZygoteProcess.java       |  8 +--
 .../android/webkit/WebViewLibraryLoader.java  |  6 +-
 core/java/android/webkit/WebViewZygote.java   |  2 +-
 .../internal/content/NativeLibraryHelper.java | 53 +++++++++++++-----
 .../com/android/server/PinnerService.java     |  4 +-
 .../am/ActivityManagerShellCommand.java       |  2 +-
 .../java/com/android/server/pm/Installer.java |  2 +-
 .../android/server/pm/InstructionSets.java    |  8 +--
 .../server/pm/PackageAbiHelperImpl.java       | 55 +++++++++++--------
 .../server/pm/PackageManagerService.java      |  2 +-
 .../server/pm/PackageManagerServiceUtils.java |  2 +-
 .../server/pm/PackageManagerShellCommand.java |  2 +-
 .../server/pm/dex/PackageDexUsage.java        |  2 +-
 .../java/com/android/server/SystemServer.java |  2 +-
 17 files changed, 123 insertions(+), 66 deletions(-)

diff --git a/cmds/am/src/com/android/commands/am/Instrument.java b/cmds/am/src/com/android/commands/am/Instrument.java
index 1e72ddf8ecfc..01d6737beeab 100644
--- a/cmds/am/src/com/android/commands/am/Instrument.java
+++ b/cmds/am/src/com/android/commands/am/Instrument.java
@@ -491,7 +491,7 @@ public class Instrument {
 
             // Choose an ABI if necessary
             if (abi != null) {
-                final String[] supportedAbis = Build.SUPPORTED_ABIS;
+                final String[] supportedAbis = Build.INTERNAL_SUPPORTED_ABIS;
                 boolean matched = false;
                 for (String supportedAbi : supportedAbis) {
                     if (supportedAbi.equals(abi)) {
diff --git a/core/java/android/os/AppZygote.java b/core/java/android/os/AppZygote.java
index 74b814ea4159..55f1dc451dfc 100644
--- a/core/java/android/os/AppZygote.java
+++ b/core/java/android/os/AppZygote.java
@@ -102,7 +102,7 @@ public class AppZygote {
     @GuardedBy("mLock")
     private void connectToZygoteIfNeededLocked() {
         String abi = mAppInfo.primaryCpuAbi != null ? mAppInfo.primaryCpuAbi :
-                Build.SUPPORTED_ABIS[0];
+                Build.INTERNAL_SUPPORTED_ABIS[0];
         try {
             mZygote = Process.ZYGOTE_PROCESS.startChildZygote(
                     "com.android.internal.os.AppZygoteInit",
diff --git a/core/java/android/os/Build.java b/core/java/android/os/Build.java
index 6bf394dc347b..7ead13038b7e 100755
--- a/core/java/android/os/Build.java
+++ b/core/java/android/os/Build.java
@@ -217,7 +217,8 @@ public class Build {
      *
      * See {@link #SUPPORTED_32_BIT_ABIS} and {@link #SUPPORTED_64_BIT_ABIS}.
      */
-    public static final String[] SUPPORTED_ABIS = getStringList("ro.product.cpu.abilist", ",");
+    public static final String[] SUPPORTED_ABIS = getStringList(
+            "ro.floral.nb.pubabi", ",", "ro.product.cpu.abilist");
 
     /**
      * An ordered list of <b>32 bit</b> ABIs supported by this device. The most preferred ABI
@@ -225,8 +226,8 @@ public class Build {
      *
      * See {@link #SUPPORTED_ABIS} and {@link #SUPPORTED_64_BIT_ABIS}.
      */
-    public static final String[] SUPPORTED_32_BIT_ABIS =
-            getStringList("ro.product.cpu.abilist32", ",");
+    public static final String[] SUPPORTED_32_BIT_ABIS = getStringList(
+            "ro.floral.nb.pubabi32", ",", "ro.product.cpu.abilist32");
 
     /**
      * An ordered list of <b>64 bit</b> ABIs supported by this device. The most preferred ABI
@@ -234,8 +235,20 @@ public class Build {
      *
      * See {@link #SUPPORTED_ABIS} and {@link #SUPPORTED_32_BIT_ABIS}.
      */
-    public static final String[] SUPPORTED_64_BIT_ABIS =
-            getStringList("ro.product.cpu.abilist64", ",");
+    public static final String[] SUPPORTED_64_BIT_ABIS = getStringList(
+            "ro.floral.nb.pubabi64", ",", "ro.product.cpu.abilist64");
+
+    /** @hide ABI order used by PackageManager, zygote, and native loaders. */
+    public static final String[] INTERNAL_SUPPORTED_ABIS = getStringList(
+            "ro.floral.nb.loadabi", ",", "ro.product.cpu.abilist");
+
+    /** @hide 32-bit ABI order used by framework internals. */
+    public static final String[] INTERNAL_SUPPORTED_32_BIT_ABIS = getStringList(
+            "ro.floral.nb.loadabi32", ",", "ro.product.cpu.abilist32");
+
+    /** @hide 64-bit ABI order used by framework internals. */
+    public static final String[] INTERNAL_SUPPORTED_64_BIT_ABIS = getStringList(
+            "ro.floral.nb.loadabi64", ",", "ro.product.cpu.abilist64");
 
     /** {@hide} */
     @TestApi
@@ -1443,6 +1456,18 @@ public class Build {
         }
     }
 
+    private static String[] getStringList(String property, String separator, String fallback) {
+        String value = SystemProperties.get(property);
+        if (value.isEmpty()) {
+            value = SystemProperties.get(fallback);
+        }
+        if (value.isEmpty()) {
+            return new String[0];
+        } else {
+            return value.split(separator);
+        }
+    }
+
     @UnsupportedAppUsage
     private static long getLong(String property) {
         try {
diff --git a/core/java/android/os/ZygoteProcess.java b/core/java/android/os/ZygoteProcess.java
index bf2898137967..682f9970ea1a 100644
--- a/core/java/android/os/ZygoteProcess.java
+++ b/core/java/android/os/ZygoteProcess.java
@@ -892,11 +892,11 @@ public class ZygoteProcess {
      */
     public void bootCompleted() {
         // Notify both the 32-bit and 64-bit zygote.
-        if (Build.SUPPORTED_32_BIT_ABIS.length > 0) {
-            bootCompleted(Build.SUPPORTED_32_BIT_ABIS[0]);
+        if (Build.INTERNAL_SUPPORTED_32_BIT_ABIS.length > 0) {
+            bootCompleted(Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0]);
         }
-        if (Build.SUPPORTED_64_BIT_ABIS.length > 0) {
-            bootCompleted(Build.SUPPORTED_64_BIT_ABIS[0]);
+        if (Build.INTERNAL_SUPPORTED_64_BIT_ABIS.length > 0) {
+            bootCompleted(Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0]);
         }
     }
 
diff --git a/core/java/android/webkit/WebViewLibraryLoader.java b/core/java/android/webkit/WebViewLibraryLoader.java
index be49fc434c79..a1e46f7c36fa 100644
--- a/core/java/android/webkit/WebViewLibraryLoader.java
+++ b/core/java/android/webkit/WebViewLibraryLoader.java
@@ -106,7 +106,7 @@ public class WebViewLibraryLoader {
     static void createRelroFile(final boolean is64Bit, @NonNull String packageName,
             @NonNull String libraryFileName) {
         final String abi =
-                is64Bit ? Build.SUPPORTED_64_BIT_ABIS[0] : Build.SUPPORTED_32_BIT_ABIS[0];
+                is64Bit ? Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0] : Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0];
 
         // crashHandler is invoked by the ActivityManagerService when the isolated process crashes.
         Runnable crashHandler = new Runnable() {
@@ -159,13 +159,13 @@ public class WebViewLibraryLoader {
         if (DEBUG) Log.v(LOGTAG, "creating relro files");
         int numRelros = 0;
 
-        if (Build.SUPPORTED_32_BIT_ABIS.length > 0) {
+        if (Build.INTERNAL_SUPPORTED_32_BIT_ABIS.length > 0) {
             if (DEBUG) Log.v(LOGTAG, "Create 32 bit relro");
             createRelroFile(false /* is64Bit */, packageName, libraryFileName);
             numRelros++;
         }
 
-        if (Build.SUPPORTED_64_BIT_ABIS.length > 0) {
+        if (Build.INTERNAL_SUPPORTED_64_BIT_ABIS.length > 0) {
             if (DEBUG) Log.v(LOGTAG, "Create 64 bit relro");
             createRelroFile(true /* is64Bit */, packageName, libraryFileName);
             numRelros++;
diff --git a/core/java/android/webkit/WebViewZygote.java b/core/java/android/webkit/WebViewZygote.java
index 2bfbe4bdfba7..f3d068ff365a 100644
--- a/core/java/android/webkit/WebViewZygote.java
+++ b/core/java/android/webkit/WebViewZygote.java
@@ -136,7 +136,7 @@ public class WebViewZygote {
                     0,  // runtimeFlags
                     "webview_zygote",  // seInfo
                     abi,  // abi
-                    TextUtils.join(",", Build.SUPPORTED_ABIS),
+                    TextUtils.join(",", Build.INTERNAL_SUPPORTED_ABIS),
                     null, // instructionSet
                     Process.FIRST_ISOLATED_UID,
                     Integer.MAX_VALUE); // TODO(b/123615476) deal with user-id ranges properly
diff --git a/core/java/com/android/internal/content/NativeLibraryHelper.java b/core/java/com/android/internal/content/NativeLibraryHelper.java
index 10750b695a39..10fb82fac225 100644
--- a/core/java/com/android/internal/content/NativeLibraryHelper.java
+++ b/core/java/com/android/internal/content/NativeLibraryHelper.java
@@ -237,6 +237,17 @@ public class NativeLibraryHelper {
         return finalRes;
     }
 
+    /**
+     * Returns whether the package contains native code for the public application ABI view.
+     *
+     * Floral keeps host ABIs in the internal loader view, but a package that contains any public
+     * guest ABI must stay on that ABI family. This prevents an auxiliary host-ABI library from
+     * changing the process ABI selected for an otherwise ARM application.
+     */
+    public static boolean hasPublicNativeLibraries(Handle handle) {
+        return findSupportedAbi(handle, Build.SUPPORTED_ABIS) >= 0;
+    }
+
     private native static int nativeFindSupportedAbi(long handle, String[] supportedAbis,
             boolean debuggable);
 
@@ -373,6 +384,13 @@ public class NativeLibraryHelper {
     public static int copyNativeBinariesWithOverride(Handle handle, File libraryRoot,
             String abiOverride, boolean isIncremental) {
         try {
+            final boolean usePublicAbis = hasPublicNativeLibraries(handle);
+            final String[] supportedAbis = usePublicAbis
+                    ? Build.SUPPORTED_ABIS : Build.INTERNAL_SUPPORTED_ABIS;
+            final String[] supported32BitAbis = usePublicAbis
+                    ? Build.SUPPORTED_32_BIT_ABIS : Build.INTERNAL_SUPPORTED_32_BIT_ABIS;
+            final String[] supported64BitAbis = usePublicAbis
+                    ? Build.SUPPORTED_64_BIT_ABIS : Build.INTERNAL_SUPPORTED_64_BIT_ABIS;
             if (handle.multiArch) {
                 // Warn if we've set an abiOverride for multi-lib packages..
                 // By definition, we need to copy both 32 and 64 bit libraries for
@@ -382,9 +400,9 @@ public class NativeLibraryHelper {
                 }
 
                 int copyRet = PackageManager.NO_NATIVE_LIBRARIES;
-                if (Build.SUPPORTED_32_BIT_ABIS.length > 0) {
+                if (supported32BitAbis.length > 0) {
                     copyRet = copyNativeBinariesForSupportedAbi(handle, libraryRoot,
-                            Build.SUPPORTED_32_BIT_ABIS, true /* use isa specific subdirs */,
+                            supported32BitAbis, true /* use isa specific subdirs */,
                             isIncremental);
                     if (copyRet < 0 && copyRet != PackageManager.NO_NATIVE_LIBRARIES &&
                             copyRet != PackageManager.INSTALL_FAILED_NO_MATCHING_ABIS) {
@@ -393,9 +411,9 @@ public class NativeLibraryHelper {
                     }
                 }
 
-                if (Build.SUPPORTED_64_BIT_ABIS.length > 0) {
+                if (supported64BitAbis.length > 0) {
                     copyRet = copyNativeBinariesForSupportedAbi(handle, libraryRoot,
-                            Build.SUPPORTED_64_BIT_ABIS, true /* use isa specific subdirs */,
+                            supported64BitAbis, true /* use isa specific subdirs */,
                             isIncremental);
                     if (copyRet < 0 && copyRet != PackageManager.NO_NATIVE_LIBRARIES &&
                             copyRet != PackageManager.INSTALL_FAILED_NO_MATCHING_ABIS) {
@@ -412,10 +430,10 @@ public class NativeLibraryHelper {
                 }
 
                 String[] abiList = (cpuAbiOverride != null) ?
-                        new String[] { cpuAbiOverride } : Build.SUPPORTED_ABIS;
-                if (Build.SUPPORTED_64_BIT_ABIS.length > 0 && cpuAbiOverride == null &&
+                        new String[] { cpuAbiOverride } : supportedAbis;
+                if (supported64BitAbis.length > 0 && cpuAbiOverride == null &&
                         hasRenderscriptBitcode(handle)) {
-                    abiList = Build.SUPPORTED_32_BIT_ABIS;
+                    abiList = supported32BitAbis;
                 }
 
                 int copyRet = copyNativeBinariesForSupportedAbi(handle, libraryRoot, abiList,
@@ -436,6 +454,13 @@ public class NativeLibraryHelper {
     public static long sumNativeBinariesWithOverride(Handle handle, String abiOverride)
             throws IOException {
         long sum = 0;
+        final boolean usePublicAbis = hasPublicNativeLibraries(handle);
+        final String[] supportedAbis = usePublicAbis
+                ? Build.SUPPORTED_ABIS : Build.INTERNAL_SUPPORTED_ABIS;
+        final String[] supported32BitAbis = usePublicAbis
+                ? Build.SUPPORTED_32_BIT_ABIS : Build.INTERNAL_SUPPORTED_32_BIT_ABIS;
+        final String[] supported64BitAbis = usePublicAbis
+                ? Build.SUPPORTED_64_BIT_ABIS : Build.INTERNAL_SUPPORTED_64_BIT_ABIS;
         if (handle.multiArch) {
             // Warn if we've set an abiOverride for multi-lib packages..
             // By definition, we need to copy both 32 and 64 bit libraries for
@@ -444,12 +469,12 @@ public class NativeLibraryHelper {
                 Slog.w(TAG, "Ignoring abiOverride for multi arch application.");
             }
 
-            if (Build.SUPPORTED_32_BIT_ABIS.length > 0) {
-                sum += sumNativeBinariesForSupportedAbi(handle, Build.SUPPORTED_32_BIT_ABIS);
+            if (supported32BitAbis.length > 0) {
+                sum += sumNativeBinariesForSupportedAbi(handle, supported32BitAbis);
             }
 
-            if (Build.SUPPORTED_64_BIT_ABIS.length > 0) {
-                sum += sumNativeBinariesForSupportedAbi(handle, Build.SUPPORTED_64_BIT_ABIS);
+            if (supported64BitAbis.length > 0) {
+                sum += sumNativeBinariesForSupportedAbi(handle, supported64BitAbis);
             }
         } else {
             String cpuAbiOverride = null;
@@ -460,10 +485,10 @@ public class NativeLibraryHelper {
             }
 
             String[] abiList = (cpuAbiOverride != null) ?
-                    new String[] { cpuAbiOverride } : Build.SUPPORTED_ABIS;
-            if (Build.SUPPORTED_64_BIT_ABIS.length > 0 && cpuAbiOverride == null &&
+                    new String[] { cpuAbiOverride } : supportedAbis;
+            if (supported64BitAbis.length > 0 && cpuAbiOverride == null &&
                     hasRenderscriptBitcode(handle)) {
-                abiList = Build.SUPPORTED_32_BIT_ABIS;
+                abiList = supported32BitAbis;
             }
 
             sum += sumNativeBinariesForSupportedAbi(handle, abiList);
diff --git a/services/core/java/com/android/server/PinnerService.java b/services/core/java/com/android/server/PinnerService.java
index 8e53101b0eab..5c8f0ae6edde 100644
--- a/services/core/java/com/android/server/PinnerService.java
+++ b/services/core/java/com/android/server/PinnerService.java
@@ -282,7 +282,7 @@ public final class PinnerService extends SystemService {
             }
             if (fileToPin.endsWith(".jar") | fileToPin.endsWith(".apk")) {
                 // Check whether the runtime has compilation artifacts to pin.
-                String arch = VMRuntime.getInstructionSet(Build.SUPPORTED_ABIS[0]);
+                String arch = VMRuntime.getInstructionSet(Build.INTERNAL_SUPPORTED_ABIS[0]);
                 String[] files = null;
                 try {
                     files = DexFile.getDexFileOutputPaths(fileToPin, arch);
@@ -723,7 +723,7 @@ public final class PinnerService extends SystemService {
 
         // determine the ABI from either ApplicationInfo or Build
         String abi = appInfo.primaryCpuAbi != null ? appInfo.primaryCpuAbi :
-                Build.SUPPORTED_ABIS[0];
+                Build.INTERNAL_SUPPORTED_ABIS[0];
         String arch = VMRuntime.getInstructionSet(abi);
         // get the path to the odex or oat file
         String baseCodePath = appInfo.getBaseCodePath();
diff --git a/services/core/java/com/android/server/am/ActivityManagerShellCommand.java b/services/core/java/com/android/server/am/ActivityManagerShellCommand.java
index 685d606f8d41..8408976cf504 100644
--- a/services/core/java/com/android/server/am/ActivityManagerShellCommand.java
+++ b/services/core/java/com/android/server/am/ActivityManagerShellCommand.java
@@ -2393,7 +2393,7 @@ final class ActivityManagerShellCommand extends ShellCommand {
             proto.flush();
         } else {
             pw.println("config: " + Configuration.resourceQualifierString(config, metrics));
-            pw.println("abi: " + TextUtils.join(",", Build.SUPPORTED_ABIS));
+            pw.println("abi: " + TextUtils.join(",", Build.INTERNAL_SUPPORTED_ABIS));
             if (inclDevice) {
                 writeDeviceConfig(null, -1, pw, config, metrics);
             }
diff --git a/services/core/java/com/android/server/pm/Installer.java b/services/core/java/com/android/server/pm/Installer.java
index 8fd545f271f3..bcb42f0b6892 100644
--- a/services/core/java/com/android/server/pm/Installer.java
+++ b/services/core/java/com/android/server/pm/Installer.java
@@ -925,7 +925,7 @@ public class Installer extends SystemService {
 
     private static void assertValidInstructionSet(String instructionSet)
             throws InstallerException {
-        for (String abi : Build.SUPPORTED_ABIS) {
+        for (String abi : Build.INTERNAL_SUPPORTED_ABIS) {
             if (VMRuntime.getInstructionSet(abi).equals(instructionSet)) {
                 return;
             }
diff --git a/services/core/java/com/android/server/pm/InstructionSets.java b/services/core/java/com/android/server/pm/InstructionSets.java
index 2d42107bdd63..d61753684c47 100644
--- a/services/core/java/com/android/server/pm/InstructionSets.java
+++ b/services/core/java/com/android/server/pm/InstructionSets.java
@@ -36,7 +36,7 @@ import java.util.List;
  */
 public class InstructionSets {
     private static final String PREFERRED_INSTRUCTION_SET =
-            VMRuntime.getInstructionSet(Build.SUPPORTED_ABIS[0]);
+            VMRuntime.getInstructionSet(Build.INTERNAL_SUPPORTED_ABIS[0]);
 
     public static String[] getAppDexInstructionSets(String primaryCpuAbi, String secondaryCpuAbi) {
         if (primaryCpuAbi != null) {
@@ -79,16 +79,16 @@ public class InstructionSets {
      * Returns deduplicated list of supported instructions for dex code.
      */
     public static String[] getAllDexCodeInstructionSets() {
-        String[] supportedInstructionSets = new String[Build.SUPPORTED_ABIS.length];
+        String[] supportedInstructionSets = new String[Build.INTERNAL_SUPPORTED_ABIS.length];
         for (int i = 0; i < supportedInstructionSets.length; i++) {
-            String abi = Build.SUPPORTED_ABIS[i];
+            String abi = Build.INTERNAL_SUPPORTED_ABIS[i];
             supportedInstructionSets[i] = VMRuntime.getInstructionSet(abi);
         }
         return getDexCodeInstructionSets(supportedInstructionSets);
     }
 
     public static List<String> getAllInstructionSets() {
-        final String[] allAbis = Build.SUPPORTED_ABIS;
+        final String[] allAbis = Build.INTERNAL_SUPPORTED_ABIS;
         final List<String> allInstructionSets = new ArrayList<String>(allAbis.length);
 
         for (String abi : allAbis) {
diff --git a/services/core/java/com/android/server/pm/PackageAbiHelperImpl.java b/services/core/java/com/android/server/pm/PackageAbiHelperImpl.java
index d851e6ced91b..9fe7c2cf4788 100644
--- a/services/core/java/com/android/server/pm/PackageAbiHelperImpl.java
+++ b/services/core/java/com/android/server/pm/PackageAbiHelperImpl.java
@@ -239,16 +239,16 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
         } else {
             // Cluster install
             final File rootDir = new File(codeFile, LIB_DIR_NAME);
-            if (!ArrayUtils.isEmpty(Build.SUPPORTED_64_BIT_ABIS)
-                    && !TextUtils.isEmpty(Build.SUPPORTED_64_BIT_ABIS[0])) {
-                final String isa = VMRuntime.getInstructionSet(Build.SUPPORTED_64_BIT_ABIS[0]);
+            if (!ArrayUtils.isEmpty(Build.INTERNAL_SUPPORTED_64_BIT_ABIS)
+                    && !TextUtils.isEmpty(Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0])) {
+                final String isa = VMRuntime.getInstructionSet(Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0]);
                 has64BitLibs = (new File(rootDir, isa)).exists();
             } else {
                 has64BitLibs = false;
             }
-            if (!ArrayUtils.isEmpty(Build.SUPPORTED_32_BIT_ABIS)
-                    && !TextUtils.isEmpty(Build.SUPPORTED_32_BIT_ABIS[0])) {
-                final String isa = VMRuntime.getInstructionSet(Build.SUPPORTED_32_BIT_ABIS[0]);
+            if (!ArrayUtils.isEmpty(Build.INTERNAL_SUPPORTED_32_BIT_ABIS)
+                    && !TextUtils.isEmpty(Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0])) {
+                final String isa = VMRuntime.getInstructionSet(Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0]);
                 has32BitLibs = (new File(rootDir, isa)).exists();
             } else {
                 has32BitLibs = false;
@@ -260,13 +260,13 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
             // ABI should be 64 bit. We can safely assume here that the bundled
             // native libraries correspond to the most preferred ABI in the list.
 
-            primaryCpuAbi = Build.SUPPORTED_64_BIT_ABIS[0];
+            primaryCpuAbi = Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0];
             secondaryCpuAbi = null;
         } else if (has32BitLibs && !has64BitLibs) {
             // The package has 32 bit libs but not 64 bit libs. Its primary
             // ABI should be 32 bit.
 
-            primaryCpuAbi = Build.SUPPORTED_32_BIT_ABIS[0];
+            primaryCpuAbi = Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0];
             secondaryCpuAbi = null;
         } else if (has32BitLibs && has64BitLibs) {
             // The application has both 64 and 32 bit bundled libraries. We check
@@ -282,11 +282,11 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
             }
 
             if (VMRuntime.is64BitInstructionSet(getPreferredInstructionSet())) {
-                primaryCpuAbi = Build.SUPPORTED_64_BIT_ABIS[0];
-                secondaryCpuAbi = Build.SUPPORTED_32_BIT_ABIS[0];
+                primaryCpuAbi = Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0];
+                secondaryCpuAbi = Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0];
             } else {
-                primaryCpuAbi = Build.SUPPORTED_32_BIT_ABIS[0];
-                secondaryCpuAbi = Build.SUPPORTED_64_BIT_ABIS[0];
+                primaryCpuAbi = Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0];
+                secondaryCpuAbi = Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0];
             }
         } else {
             primaryCpuAbi = null;
@@ -321,6 +321,13 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
         NativeLibraryHelper.Handle handle = null;
         try {
             handle = AndroidPackageUtils.createNativeLibraryHandle(pkg);
+            final boolean usePublicAbis = NativeLibraryHelper.hasPublicNativeLibraries(handle);
+            final String[] supportedAbis = usePublicAbis
+                    ? Build.SUPPORTED_ABIS : Build.INTERNAL_SUPPORTED_ABIS;
+            final String[] supported32BitAbis = usePublicAbis
+                    ? Build.SUPPORTED_32_BIT_ABIS : Build.INTERNAL_SUPPORTED_32_BIT_ABIS;
+            final String[] supported64BitAbis = usePublicAbis
+                    ? Build.SUPPORTED_64_BIT_ABIS : Build.INTERNAL_SUPPORTED_64_BIT_ABIS;
             // TODO(multiArch): This can be null for apps that didn't go through the
             // usual installation process. We can calculate it again, like we
             // do during install time.
@@ -335,16 +342,16 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
             if (pkg.isMultiArch()) {
                 int abi32 = PackageManager.NO_NATIVE_LIBRARIES;
                 int abi64 = PackageManager.NO_NATIVE_LIBRARIES;
-                if (Build.SUPPORTED_32_BIT_ABIS.length > 0) {
+                if (supported32BitAbis.length > 0) {
                     if (extractLibs) {
                         Trace.traceBegin(TRACE_TAG_PACKAGE_MANAGER, "copyNativeBinaries");
                         abi32 = NativeLibraryHelper.copyNativeBinariesForSupportedAbi(handle,
-                                nativeLibraryRoot, Build.SUPPORTED_32_BIT_ABIS,
+                                nativeLibraryRoot, supported32BitAbis,
                                 useIsaSpecificSubdirs, onIncremental);
                     } else {
                         Trace.traceBegin(TRACE_TAG_PACKAGE_MANAGER, "findSupportedAbi");
                         abi32 = NativeLibraryHelper.findSupportedAbi(
-                                handle, Build.SUPPORTED_32_BIT_ABIS);
+                                handle, supported32BitAbis);
                     }
                     Trace.traceEnd(TRACE_TAG_PACKAGE_MANAGER);
                 }
@@ -358,16 +365,16 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
                 maybeThrowExceptionForMultiArchCopy(
                         "Error unpackaging 32 bit native libs for multiarch app.", abi32);
 
-                if (Build.SUPPORTED_64_BIT_ABIS.length > 0) {
+                if (supported64BitAbis.length > 0) {
                     if (extractLibs) {
                         Trace.traceBegin(TRACE_TAG_PACKAGE_MANAGER, "copyNativeBinaries");
                         abi64 = NativeLibraryHelper.copyNativeBinariesForSupportedAbi(handle,
-                                nativeLibraryRoot, Build.SUPPORTED_64_BIT_ABIS,
+                                nativeLibraryRoot, supported64BitAbis,
                                 useIsaSpecificSubdirs, onIncremental);
                     } else {
                         Trace.traceBegin(TRACE_TAG_PACKAGE_MANAGER, "findSupportedAbi");
                         abi64 = NativeLibraryHelper.findSupportedAbi(
-                                handle, Build.SUPPORTED_64_BIT_ABIS);
+                                handle, supported64BitAbis);
                     }
                     Trace.traceEnd(TRACE_TAG_PACKAGE_MANAGER);
                 }
@@ -381,11 +388,11 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
                         throw new PackageManagerException(INSTALL_FAILED_INTERNAL_ERROR,
                                 "Shared library native lib extraction not supported");
                     }
-                    primaryCpuAbi = Build.SUPPORTED_64_BIT_ABIS[abi64];
+                    primaryCpuAbi = supported64BitAbis[abi64];
                 }
 
                 if (abi32 >= 0) {
-                    final String abi = Build.SUPPORTED_32_BIT_ABIS[abi32];
+                    final String abi = supported32BitAbis[abi32];
                     if (abi64 >= 0) {
                         if (pkg.isUse32BitAbi()) {
                             secondaryCpuAbi = primaryCpuAbi;
@@ -399,7 +406,7 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
                 }
             } else {
                 String[] abiList = (cpuAbiOverride != null)
-                        ? new String[]{cpuAbiOverride} : Build.SUPPORTED_ABIS;
+                        ? new String[]{cpuAbiOverride} : supportedAbis;
 
                 // If an app that contains RenderScript has target API level < 21, it needs to run
                 // with 32-bit ABI, and its APK file will contain a ".bc" file.
@@ -411,10 +418,10 @@ final class PackageAbiHelperImpl implements PackageAbiHelper {
                 // we will fail the installation for such an app because it won't be able to run.
                 boolean needsRenderScriptOverride = false;
                 // No need to check if the device only supports 32-bit
-                if (Build.SUPPORTED_64_BIT_ABIS.length > 0 && cpuAbiOverride == null
+                if (supported64BitAbis.length > 0 && cpuAbiOverride == null
                         && NativeLibraryHelper.hasRenderscriptBitcode(handle)) {
-                    if (Build.SUPPORTED_32_BIT_ABIS.length > 0) {
-                        abiList = Build.SUPPORTED_32_BIT_ABIS;
+                    if (supported32BitAbis.length > 0) {
+                        abiList = supported32BitAbis;
                         needsRenderScriptOverride = true;
                     } else {
                         throw new PackageManagerException(
diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index 6da70ad77d67..26ad7aaf3170 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -14597,7 +14597,7 @@ public class PackageManagerService extends IPackageManager.Stub
         // the same UID correctly.
         if (isPlatformPackage) {
             parsedPackage.setPrimaryCpuAbi(VMRuntime.getRuntime().is64Bit() ?
-                    Build.SUPPORTED_64_BIT_ABIS[0] : Build.SUPPORTED_32_BIT_ABIS[0]);
+                    Build.INTERNAL_SUPPORTED_64_BIT_ABIS[0] : Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0]);
         }
 
         // If there's a mismatch between the abi-override in the package setting
diff --git a/services/core/java/com/android/server/pm/PackageManagerServiceUtils.java b/services/core/java/com/android/server/pm/PackageManagerServiceUtils.java
index 8b5abf3afe51..e11b4cd8bb24 100644
--- a/services/core/java/com/android/server/pm/PackageManagerServiceUtils.java
+++ b/services/core/java/com/android/server/pm/PackageManagerServiceUtils.java
@@ -320,7 +320,7 @@ public class PackageManagerServiceUtils {
      * the running device.
      */
     public static boolean checkISA(String isa) {
-        for (String abi : Build.SUPPORTED_ABIS) {
+        for (String abi : Build.INTERNAL_SUPPORTED_ABIS) {
             if (VMRuntime.getInstructionSet(abi).equals(isa)) {
                 return true;
             }
diff --git a/services/core/java/com/android/server/pm/PackageManagerShellCommand.java b/services/core/java/com/android/server/pm/PackageManagerShellCommand.java
index 1aa80a953d6c..7d39a0d040cc 100644
--- a/services/core/java/com/android/server/pm/PackageManagerShellCommand.java
+++ b/services/core/java/com/android/server/pm/PackageManagerShellCommand.java
@@ -3115,7 +3115,7 @@ class PackageManagerShellCommand extends ShellCommand {
             return abi;
         }
 
-        final String[] supportedAbis = Build.SUPPORTED_ABIS;
+        final String[] supportedAbis = Build.INTERNAL_SUPPORTED_ABIS;
         for (String supportedAbi : supportedAbis) {
             if (supportedAbi.equals(abi)) {
                 return abi;
diff --git a/services/core/java/com/android/server/pm/dex/PackageDexUsage.java b/services/core/java/com/android/server/pm/dex/PackageDexUsage.java
index 3d63b75c5da1..9fcf4f0395f4 100644
--- a/services/core/java/com/android/server/pm/dex/PackageDexUsage.java
+++ b/services/core/java/com/android/server/pm/dex/PackageDexUsage.java
@@ -346,7 +346,7 @@ public class PackageDexUsage extends AbstractStatsBase<Void> {
         PackageUseInfo currentPackageData = null;
 
         Set<String> supportedIsas = new HashSet<>();
-        for (String abi : Build.SUPPORTED_ABIS) {
+        for (String abi : Build.INTERNAL_SUPPORTED_ABIS) {
             supportedIsas.add(VMRuntime.getInstructionSet(abi));
         }
         while ((line = in.readLine()) != null) {
diff --git a/services/java/com/android/server/SystemServer.java b/services/java/com/android/server/SystemServer.java
index 33e32393090e..d8476134e820 100644
--- a/services/java/com/android/server/SystemServer.java
+++ b/services/java/com/android/server/SystemServer.java
@@ -1380,7 +1380,7 @@ public final class SystemServer implements Dumpable {
                     Slog.i(TAG, SECONDARY_ZYGOTE_PRELOAD);
                     TimingsTraceAndSlog traceLog = TimingsTraceAndSlog.newAsyncLog();
                     traceLog.traceBegin(SECONDARY_ZYGOTE_PRELOAD);
-                    if (!Process.ZYGOTE_PROCESS.preloadDefault(Build.SUPPORTED_32_BIT_ABIS[0])) {
+                    if (!Process.ZYGOTE_PROCESS.preloadDefault(Build.INTERNAL_SUPPORTED_32_BIT_ABIS[0])) {
                         Slog.e(TAG, "Unable to preload default resources");
                     }
                     traceLog.traceEnd();
